java/arcs/core/data/proto/ParticleProtoDecoder.kt

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data.proto
+
+import arcs.core.data.ParticleSpec
+import arcs.core.data.Recipe.Handle
+import arcs.core.data.Recipe.Particle
+import arcs.core.data.Recipe.Particle.HandleConnection
+
+/** Class contains information that is required when decoding particle protos.
+ *
+ * @property particleSpecs [ParticleSpec] instances in the context indexed by name.
+ * @property recipeHandles [Handle] instances in a [Recipe] indexed by name.
+ */
+class DecodingContext(
+    var particleSpecs: Map<String, ParticleSpec>,
+    var recipeHandles: Map<String, Handle>
+)
+
+/** Converts a [HandleConnectionProto] into a [Handle]. */
+fun HandleConnectionProto.decode(
+    particleSpec: ParticleSpec,
+    context: DecodingContext
+): HandleConnection {
+    val handleSpec = particleSpec.connections[name]
+    requireNotNull(handleSpec) {
+        "HandleConnection '$name' not found in ParticleSpec '${particleSpec.name}'."
+    }
+    val recipeHandle = context.recipeHandles[handle]
+    requireNotNull(recipeHandle) {
+        "Handle '$handle' not found when decoding ParticleProto '${particleSpec.name}'."
+    }
+    return HandleConnection(handleSpec, recipeHandle)
```

Comment:
`requireNotNull` returns the non-nullable value of the argument, so you can do this:
```suggestion
    val handleSpec = requireNotNull(particleSpec.connections[name]) {
        "HandleConnection '$name' not found in ParticleSpec '${particleSpec.name}'."
    }
    val recipeHandle = requireNotNull(context.recipeHandles[handle]) {
        "Handle '$handle' not found when decoding ParticleProto '${particleSpec.name}'."
    }
    return HandleConnection(handleSpec, recipeHandle)
```

Reply:
TIL. Thanks!